core/src/main/java/software/amazon/awssdk/pagination/PaginatedResponsesIterable.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.pagination;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+
+public final class PaginatedResponsesIterable<ResponseT> implements SdkIterable<ResponseT> {
+
+    private final ResponseT firstResponse;
+    private final Function<ResponseT, ResponseT> getNextResponse;
+
+    public PaginatedResponsesIterable(ResponseT firstResponse,
+                                      Function<ResponseT, ResponseT> getNextResponse) {
```

Comment:
Instead of `Function` can we create an interface like `NextResponseSupplier` and annotate with `@FunctionalInterface` instead? That way we can have a more descriptive method like `nextResponse` instead of `apply`

Ditto for `PaginatedItemsIterable`.

Reply:
I am okay with this change. Will do it

Reply:
Added an interface for NextResponseSupplier. 
I am not sure if I want to do the same for getIterator in PaginatedItemsIterable class. The result_key can be a list and we need to generate multiple classes for each key in the list. This will bloat the response class for paginated APIs.